Add 3-phase current formulas

shsms opened the pull request:
This PR introduces some new types for streaming 3-phase data:
- `Sample3Phase` :: for holding a 3-phase measurement.
- `FormulaChannel3Phase/FormulaReceiver3Phase` :: composable channel/receiver types for streaming `Sample3Phase` objects.
- `FormulaEngine3Phase` :: for combining `Sample` objects from `FormulaEngine` into `Sample3Phase` objects.
- `HigherOrderFormulaBuilder3Phase` :: For composing `FormulaReceiver3Phase` and generating new `FormulaEngine3Phase` from them.

It also adds formula generators for 3-phase `grid_current` and `ev_charger_current`,  and exposes them through the `LogicalMeter`.

shsms committed:
Add a new sample type for holding 3-phase logical meter outputs

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Convert `FormulaChannel` into an ABC and specialize it for `Sample`s

This is achieved by converting the original implementations of the
following classes into Abstract Base Classes that are then derived
from the base class with the same name, for use with `Sample` and
`FormulaEngine` objects.

 - `FormulaChannel` -> `_BaseFormulaChannel` -> `FormulaChannel`
 - `FormulaReceiver` -> `_BaseFormulaReceiver` -> `FormulaReceiver`
 - `HigherOrderFormulaBuilder` -> `_BaseHOFormulaBuilder` -> `HigherOrderFormulaBuilder`

The will be specialized for 3-phase data also,  after a 3-phase formula
engine implementation is added.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Extract formula eval logic from FormulaEngine into a separate class

... called `FormulaEvaluator`, so that it can be reused in the engine
for 3-phase formulas.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add support for ev chargers and 3-phase current in mock microgrid

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add a 3-phase formula engine implementation

It is a simple implementation that takes the outputs of 3 single-phase
formula engines and packages received samples into `Sample3Phase` objects.

Also add 3-phase channel/receiver/builder specializations using the
newly introduced `FormulaEngine3Phase`.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add formula generators for 3-phase grid and ev charger current

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add `grid_current` and `ev_charger_current` methods to `LogicalMeter`

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

ela-kotulska-frequenz commented on src/frequenz/sdk/timeseries/logical_meter/_formula_engine.py:
```diff
@@ -8,10 +8,11 @@
 import asyncio
```
Would it be ok to separate this file (_formula_engine.py) into 3 files?
* _base_formula_engine.py
* _formula_engine.py
* _3Phase_formula_engine.py

Generics can stay. I am telling only about separating this file .

shsms replied:
Unfortunately,  to do this,  we'll have to do some sacrifices to type safety,  or else, there will be cyclic imports.  

Specifically,  the generic types `_GenericEngine`, etc will have to be unbounded,  because otherwise the base classes file will have to import each engine implementation separately.

So we'll have `_GenericEngine = TypeVar("_GenericEngine")` instead of `= TypeVar("_GenericEngine", FormulaEngine, FormulaEngine3Phase)`
and `ReceiverType: Type[Any]` instead of `: Type[FormulaReceiver | FormulaReceiver3Phase]`.

I think that's tighter constraints and better type checks are more valuable than smaller files.  But I guess it is a tough decision.

ela-kotulska-frequenz commented on src/frequenz/sdk/timeseries/logical_meter/_formula_engine.py:
```diff
@@ -222,6 +222,72 @@ def new_receiver(
+        await cancel_and_await(self._task)
+
+    def new_receiver(
+        self, name: Optional[str] = None, max_size: int = 50
```
why max_size: int = 50?
receiver has default max_size = 10.

shsms replied:
No,  receivers have default 50:  https://frequenz-floss.github.io/frequenz-channels-python/v0.11/reference/frequenz/channels/#frequenz.channels._broadcast.Broadcast.new_receiver

ela-kotulska-frequenz replied:
Thanks!
Then we should not override it, here. Yes?

shsms replied:
I don't know, if people want to, they can override, right?

or are you talking about changing the default? my understanding is that 50 has been a good default for our use cases.  do you think there is a reason to change it?

matthias-wende-frequenz replied:
Maybe you should store the default somewhere, like e.g. `DEFAULT_CHANNEL_SIZE` and use it everywhere.

matthias-wende-frequenz replied:
or as an alternative, check here if it has been set by the user and if not don't pass it to the Channels `new_receiver()` function.

ela-kotulska-frequenz replied:
I mean that here you override the default Receiver's max_size.
So if anytime in the future receivers default size change, then it will change anywhere, except for this place.
Why you do this?

If you need receivers default max_size, then you could do:
`def new_receiver(self, name: Optional[str] = None, max_size: Optional[int] = None) `

shsms replied:
Why should we expect that their buffer sizes should be coupled?  `FormulaReceiver`s are very specific to a single use case and with them,  we can fine tune the default more easily.

Broadcast `Receiver`s are more general purpose and we chose 50 randomly and it seems to work for most cases.  But by decoupling them,  we get the freedom to change the general purpose `Receiver`'s buffer size without affecting the `FormulaReceiver`.

shsms committed:
Update to frequenz-channels v0.12.0

frequenz-channels v0.12.0 provides a `_deactivate` method for
broadcast receivers.  This commit drops a custom implementation for
`_deactivate` in `_BaseFormulaReceiver`, so that the upstream method
gets used.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

ela-kotulska-frequenz commented on src/frequenz/sdk/timeseries/logical_meter/_formula_generators/_grid_current_formula.py:
```diff
@@ -0,0 +1,97 @@
+            ),
+        )
+
+    async def _gen_phase_formula(
+        self,
+        grid_successors: Set[Component],
+        metric_id: ComponentMetricId,
+    ) -> FormulaEngine:
+        builder = self._get_builder("grid-current", metric_id)
```
Just curious:
How is it possible that this method doesn't need description? :)

shsms replied:
our tools allow for private methods to not have a docstring. :-)

ela-kotulska-frequenz commented on src/frequenz/sdk/timeseries/logical_meter/_formula_generators/_ev_charger_current_formula.py:
```diff
@@ -0,0 +1,91 @@
+    ) -> FormulaEngine:
+        builder = self._get_builder("ev-current", metric_id)
+
+        # generate a formula that just adds values from all commponents that are
```
Comment seems to be wrong.
`generate a formula that just adds values from the ev_chargers`
typo `commponents`

shsms replied:
fixed.

shsms commented:
Rebased to latest head.
